### ISSUE TITLE
perf(deepseek-v3): W8A16 fused-MoE — v6e-64 epmoe→fused +62% (3291→5315), v7x-16 +51%

### ISSUE THREAD

**Issue by zorrofox** [sgl-project/sglang-jax#1318]
## Summary

DeepSeek-V3 671B (official FP8 block-wise checkpoint, W8A16 inference) on TPU currently runs only via `--moe-backend epmoe`. xprof shows `EPMoE.__call__` does `reshard(hidden_states, P(None))` per layer (replicate-then-mask), which all-gathers `[T, 7168]` × 58 layers — **40% of EXTEND step** at T=8192. `FusedEPMoE` avoids this but does not support static block-wise scales.

After wiring static block-scale into the fused path + tune entries (benchmark: `bench_serving 1K/1K --random-range-ratio 1.0`, sglang-jax docs convention):

| | epmoe (main@8abb3317 + tune entries) | **fused (this work)** | Δ |
|---|---|---|---|
| **v6e-64** cc=420 | 2958 | **5299** | **+79%** |
| **v6e-64** cc=512 | 3291 | **5315** | **+62%** |
| **v7x-16** cc=512 | 2442¹ | **3696** | **+51%** |
| ITL P50 (v6e@512) | 87.2 ms | **57.5 ms** | −34% |
| GSM8K 200 | — | 0.96 / 0.965 | ✅ |

¹ v7x baseline = fused without v7x tune entries (epmoe not measured on v7x).

> **Caliber note**: an earlier revision used `--random-range-ratio 0.0` for
> v6e (in/out len ∈ U[1,1000]), which inflated the relative gain to +192%
> due to prefill-first scheduling amplifying the epmoe EXTEND bottleneck.
> All numbers above re-measured at ratio=1.0.

## Bugs found along the way

1. `blockwise_utils.get_blockwise_kernel_config`: Tier-1 fuzzy match can return `batch_block_size=1` (from #1191 GLM entries), clamped to `n_batch∈(1,8)` → violates Pallas TPU `block_m % 8 == 0 OR == array_m`, raises lowering error.
2. `weight_utils._maybe_convert_epmoe_scale_for_kernel`: `jnp.repeat` on `[E, k_blocks, n_blocks]` materializes a fully-replicated 117 MB array per expert weight → OOM on the fused path during weight load.
3. `tp_worker` `max_running_requests`: attn-backend cap is not ep-aware (e.g. 408 on v7x-16 ep=32) → fused precompile rejects `bs % ep != 0`.

## PR breakdown

- [ ] #1319 `fix(kernels): three alignment/sharding fixes for FP8 block-quant models` — standalone
- [ ] #1320 `perf(kernels): v6e+v7x MLA/quantized_matmul tune entries for DeepSeek-V3` — standalone
- [ ] #1321 `feat(fused_moe): static block-wise scale wiring for DeepSeek-V3` — standalone diff; **runtime** needs #1319 (else fused load OOM)
- [ ] #1322 `perf(fused_moe): v6e/v7x DECODE block-config entries for DeepSeek-V3` — standalone diff; entries inert until #1321 lands
- [ ] cookbook #1175 §4.2 update (after ①–④ merged)

## GLM-5.1 #1259 comparison (resolved)

Re-measured DeepSeek-V3 epmoe at `ratio=1.0` cc=420 = **2958**, consistent with GLM-5.1 #1259 = 2759 (same MoE config, same epmoe path). The earlier-reported gap (V3 epmoe = 927) was a `--random-range-ratio` mismatch (0.0 vs 1.0).

**Comment by zorrofox** [comment]
## Benchmark data corrected (re-measured at `--random-range-ratio 1.0`)

The earlier revision of this issue (and #1321/#1322) reported v6e-64 numbers
measured with `--random-range-ratio` **unset** (defaults to `0.0` →
in/out len ∈ U[1,1000]), which is **not** the sglang-jax docs convention
(`ratio=1.0`, fixed len). At `ratio=0` the prefill-first scheduler interleaves
sparse EXTEND steps that disproportionately penalize epmoe's per-layer
all-gather, inflating the relative gain. Additionally the "GLM-5.1 v7x-16=2122"
reference had no traceable source and was removed.

**Re-measured at `ratio=1.0`** (all runs: fresh server, `kubectl exec` rank0,
`1K/1K np=5×cc seed=42`, raw logs in repo `logs/v6e-{repro,epmoe-cc512-CLEAN,baseline-epmoe-cc420}.log`
and `logs/v7x-retest-*.log`):

| | epmoe | fused | Δ |
|---|---|---|---|
| v6e-64 cc=420 | 2958 | 5299 | +79% |
| v6e-64 cc=512 | 3291 | **5315** | **+62%** |
| v7x-16 cc=512 | 2442¹ | 3696 | +51% |

¹ fused without v7x tune entries (not epmoe).

**What didn't change**: the xprof root-cause (epmoe `reshard(P(None))` = 40% of
EXTEND), the code changes, and accuracy (GSM8K 0.96/0.965). Only the throughput
caliber and the relative-gain numbers were wrong. This also resolves the
"GLM #1259 = 2759 vs V3 epmoe = 927" gap — at `ratio=1.0` V3 epmoe = 2958,
consistent with GLM.

Repro:
```bash
python -m sgl_jax.bench_serving --backend sgl-jax --host 127.0.0.1 --port 30271 \
  --dataset-name random --random-input-len 1000 --random-output-len 1000 \
  --random-range-ratio 1.0 --num-prompts 2560 --max-concurrency 512 --seed 42
```

Apologies for the noise — the issue body and #1321/#1322 benchmark tables are
now updated to the corrected numbers above.
